### PR TITLE
[Crear modelo de User] User Model implentation

### PR DESCRIPTION
**Pull request by jhonmunoz**
## Summary

- User Model implementation (ManyToMany relationship to Books)
- Exception handling

## Trello Card

https://trello.com/c/J9kZfwst/19-crear-modelo-de-user

### CODE REVIEW

File: src/main/java/com/wolox/training/repositories/UserRepository.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.wolox.training.repositories;
+
+import com.wolox.training.models.Book;
+import org.springframework.data.repository.CrudRepository;
+
+public interface UserRepository extends CrudRepository<Book, Long> {
```
Comment (rodrigofranc) [inline src/main/java/com/wolox/training/repositories/UserRepository.java]: This repo is not necessary yet

File: src/main/java/com/wolox/training/models/Book.java [hunk 2]
```diff
@@ -51,6 +56,17 @@ public Book(String genre, String author, String title, String subtitle, String p
     @Column(nullable = false)
     private String isbn;
 
+    @ManyToMany(mappedBy = "books", cascade = {CascadeType.PERSIST, CascadeType.MERGE})
```
Comment (rodrigofranc) [inline src/main/java/com/wolox/training/models/Book.java]: `cascade` option is not necessary here

File: src/main/java/com/wolox/training/models/User.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    private String username;
+
+    @Column(nullable = false)
+    private String name;
+
+    @Column(nullable = false)
+    private LocalDate birthDate;
+
+    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
+    @JoinTable(name = "book_user",
```
Comment (rodrigofranc) [inline src/main/java/com/wolox/training/models/User.java]: @JoinTable is not required here, in @manytomany relationships JPA do this work for you

File: src/main/java/com/wolox/training/models/User.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        }
+    }
+
+    public void addBook(Book book) {
+        checkAlreadyOwned(book);
+        getBooks().add(book);
+    }
+
+    private void checkAlreadyOwned(final Book book){
+        if (getBooks().contains(book)){
```
Comment (rodrigofranc) [inline src/main/java/com/wolox/training/models/User.java]: Is not necessary call `getBooks()` here, directly use `this.books`

File: src/main/java/com/wolox/training/models/User.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    public void setBooks(final List<Book> books) {
+        for (Book book : books) {
+            addBook(book);
+        }
+    }
+
+    public void addBook(Book book) {
+        checkAlreadyOwned(book);
+        getBooks().add(book);
```
Comment (rodrigofranc) [inline src/main/java/com/wolox/training/models/User.java]: Same here

File: src/main/java/com/wolox/training/models/User.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    @Column(nullable = false)
+    private String username;
+
+    @Column(nullable = false)
+    private String name;
+
+    @Column(nullable = false)
+    private LocalDate birthDate;
+
+    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
```
Comment (rodrigofranc) [inline src/main/java/com/wolox/training/models/User.java]: Here the relationships need to be REFRESH, PERSIST is ok

Persist says: If you persist the Users instance to the DB, then persist every item that this collection has in the DB as well.
Merge says: If you changed the in-memory values of the collection, then merge this changes to the DB so these are reflected there.
Refresh says: If something changed in the database values of the collection, then refresh those changes in the in-memory values.

File: src/main/java/com/wolox/training/models/Book.java [hunk 7]
```diff
@@ -51,6 +56,17 @@ public Book(String genre, String author, String title, String subtitle, String p
...
+    private List<User> users;
```
Comment (rodrigofranc) [inline src/main/java/com/wolox/training/models/Book.java]: Initialize this like books in User model

**Review by rodrigofranc — CHANGES_REQUESTED** [review]

File: src/main/java/com/wolox/training/models/User.java (hunk 3 above)
Comment (jhonmunoz) [reply src/main/java/com/wolox/training/models/User.java]: nice t know

File: src/main/java/com/wolox/training/models/Book.java (hunk 2 above)
Comment (jhonmunoz) [reply src/main/java/com/wolox/training/models/Book.java]: Does it take the values from the User?

File: src/main/java/com/wolox/training/models/User.java (hunk 6 above)
Comment (rodrigofranc) [reply src/main/java/com/wolox/training/models/User.java]: sorry my mistake the the relationships need to be REFRESH y MERGE

File: src/main/java/com/wolox/training/models/Book.java (hunk 2 above)
Comment (rodrigofranc) [reply src/main/java/com/wolox/training/models/Book.java]: yes, JPA infers this 😁

**Review by rodrigofranc — APPROVED** [review]

**Pull request merged**